rust/datafusion/src/datasource/memory.rs

```diff
@@ -107,7 +107,7 @@ impl MemTable {
 
     /// Create a mem table by reading from another data source
     pub async fn load(
-        t: &dyn TableProvider,
+        t: Box<dyn TableProvider + Send + Sync>,
```

Comment:
```suggestion
        t: Arc<dyn TableProvider + Send + Sync>,
```
I suggest using `Arc` to follow the convention in ExecutionContext:
https://github.com/apache/arrow/blob/master/rust/datafusion/src/execution/context.rs#L565

I don't think there is any reason `load` needs exclusive ownership over the `TableProvider`

Reply:
Thanks @alamb .

You make a good point and it is confusing to me why `Box` is used in some places and `Arc` in others (given the I thought ideal characteristics of `Arc`): e.g.

```rust
/// Execution context for registering data sources and executing queries
#[derive(Clone)]
pub struct ExecutionContextState {
    /// Data sources that are registered with the context
    pub datasources: HashMap<String, Arc<dyn TableProvider + Send + Sync>>,
```

vs

```rust
    pub fn register_table(
        &mut self,
        name: &str,
        provider: Box<dyn TableProvider + Send + Sync>,
```

Reply:
I can take a shot at cleaning up all the APIs to take an Arc over the next few days. I think that would be the ideal outcome.

Reply:
Great. I did replace everything with Arc locally for testing and tests pass at least.